File: modules/ls-bsp/src/ls/bsp/session.scala
package ls.bsp

import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.io.OutputStream
import java.nio.charset.StandardCharsets
import java.nio.file.Path
import java.util.concurrent.CancellationException
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutionException
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.ThreadFactory
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import java.util.concurrent.atomic.AtomicInteger

import scala.concurrent.duration.*
import scala.jdk.CollectionConverters.*
import scala.util.control.NonFatal

import ch.epfl.scala.bsp4j.*
import org.eclipse.lsp4j.jsonrpc.Launcher

/** Remote interface this client speaks: core BSP plus the Scala extension
  * (buildTarget/scalacOptions and friends).
  */
trait LsBuildServer extends BuildServer with ScalaBuildServer

final case class BspSessionConfig(
    clientName: String = "scala3-bsp-semantic-ls",
    clientVersion: String = "0.1.0",
    requestTimeout: FiniteDuration = 30.seconds,
    shutdownTimeout: FiniteDuration = 5.seconds
)

/** Typed result of buildTarget/compile. */
enum BspCompileOutcome:
  case Ok(originId: Option[String])
  case Failed(statusCode: StatusCode, originId: Option[String])

  def isOk: Boolean = this match
    case Ok(_) => true
    case Failed(_, _) => false

/** One live connection to a BSP server. Every request is bounded by
  * `config.requestTimeout` and failures are rethrown as [[BspException]]
  * with a typed [[BspError]]. Not request-reentrant safeguards are needed:
  * lsp4j serializes writes internally, so the session is thread-safe.
  */
final class BspSession private[bsp] (
    val workspaceRoot: Path,
    process: Option[Process],
    input: InputStream,
    output: OutputStream,
    executor: ExecutorService,
    server: LsBuildServer,
    listening: java.util.concurrent.Future[Void],
    config: BspSessionConfig
):
  @volatile private var initializeResult: Option[InitializeBuildResult] = None
  @volatile private var closed = false

  def isClosed: Boolean = closed

  /** Alive-flag of the launched server process; None for stream-connected
    * sessions (tests, sockets).
    */
  def serverProcessAlive: Option[Boolean] = process.map(_.isAlive)

  /** Capabilities from build/initialize; None before [[initialize]] ran. */
  def serverCapabilities: Option[BuildServerCapabilities] =
    initializeResult.map(_.getCapabilities)

  /** build/initialize (languageIds = ["scala"]) followed by the
    * build/initialized notification.
    */
  def initialize(): InitializeBuildResult =
    val caps = new BuildClientCapabilities(java.util.List.of("scala"))
    val params = new InitializeBuildParams(
      config.clientName,
      config.clientVersion,
      Bsp4j.PROTOCOL_VERSION,
      workspaceRoot.toUri.toString,
      caps
    )
    val result = request("build/initialize")(_.buildInitialize(params))
    initializeResult = Some(result)
    notification("build/initialized")(_.onBuildInitialized())
    result

  def workspaceBuildTargets(): Vector[BuildTarget] =
    val result = request("workspace/buildTargets")(_.workspaceBuildTargets())
    listOf("workspace/buildTargets", result.getTargets)

  def buildTargetSources(bspIds: Vector[String]): Vector[SourcesItem] =
    val params = new SourcesParams(idsOf(bspIds))
    val result = request("buildTarget/sources")(_.buildTargetSources(params))
    listOf("buildTarget/sources", result.getItems)

  def buildTargetScalacOptions(bspIds: Vector[String]): Vector[ScalacOptionsItem] =
    val params = new ScalacOptionsParams(idsOf(bspIds))
    val result = request("buildTarget/scalacOptions")(_.buildTargetScalacOptions(params))
    listOf("buildTarget/scalacOptions", result.getItems)

  /** buildTarget/compile. Diagnostics arrive through the client handlers
    * while the request is in flight; the status code is returned typed.
    */
  def compile(bspIds: Vector[String], originId: Option[String] = None): BspCompileOutcome =
    val params = new CompileParams(idsOf(bspIds))
    originId.foreach(params.setOriginId)
    val result = request("buildTarget/compile")(_.buildTargetCompile(params))
    val resultOrigin = Option(result.getOriginId)
    result.getStatusCode match
      case StatusCode.OK => BspCompileOutcome.Ok(resultOrigin)
      case null =>
        throw BspException(BspError.InvalidResponse("buildTarget/compile", "missing statusCode"))
      case other => BspCompileOutcome.Failed(other, resultOrigin)

  /** Raw buildTarget/inverseSources call, regardless of capabilities. */
  def serverInverseSources(uri: String): Vector[String] =
    val params = new InverseSourcesParams(new TextDocumentIdentifier(uri))
    val result = request("buildTarget/inverseSources")(_.buildTargetInverseSources(params))
    listOf("buildTarget/inverseSources", result.getTargets).map(_.getUri)

  /** Uses the server when it advertises inverseSourcesProvider, otherwise
    * falls back to the local uri -> target map of the project model.
    */
  def inverseSources(uri: String, model: BspProjectModel): Vector[String] =
    val advertised = serverCapabilities.exists { caps =>
      Option(caps.getInverseSourcesProvider).exists(_.booleanValue)
    }
    if advertised then serverInverseSources(uri)
    else model.uriToTarget.get(uri).toVector

  /** Graceful buildShutdown + build/exit, then stream/process teardown. Each
    * step is best-effort and bounded by `config.shutdownTimeout`.
    */
  def shutdown(): Unit =
    if closed then return
    try
      try
        server.buildShutdown().get(config.shutdownTimeout.toMillis, TimeUnit.MILLISECONDS)
        ()
      catch
        case _: InterruptedException => Thread.currentThread().interrupt()
        case NonFatal(_) => ()
      try server.onBuildExit()
      catch case NonFatal(_) => ()
    finally close()

  /** Hard teardown: stops listening, closes streams, terminates the server
    * process (waiting `config.shutdownTimeout`, then destroy, then forcibly).
    */
  def close(): Unit =
    if closed then return
    closed = true
    listening.cancel(true)
    closeQuietly(output)
    closeQuietly(input)
    executor.shutdown()
    process.foreach(terminate)

  private def terminate(p: Process): Unit =
    if waitQuietly(p, config.shutdownTimeout.toMillis) then return
    p.destroy()
    if waitQuietly(p, 1000L) then return
    p.destroyForcibly()
    waitQuietly(p, 1000L)
    ()

  private def waitQuietly(p: Process, millis: Long): Boolean =
    try p.waitFor(millis, TimeUnit.MILLISECONDS)
    catch
      case _: InterruptedException =>
        Thread.currentThread().interrupt()
        false

  private def closeQuietly(c: java.io.Closeable): Unit =
    try c.close()
    catch case NonFatal(_) => ()

  private def idsOf(bspIds: Vector[String]): java.util.List[BuildTargetIdentifier] =
    bspIds.map(new BuildTargetIdentifier(_)).asJava

  private def listOf[A](method: String, list: java.util.List[A]): Vector[A] =
    if list == null then throw BspException(BspError.InvalidResponse(method, "missing list field"))
    list.asScala.toVector

  private def request[A](method: String)(call: LsBuildServer => CompletableFuture[A]): A =
    if closed then throw BspException(BspError.SessionClosed(method))
    val future =
      try call(server)
      catch case NonFatal(e) => throw BspException(BspError.RequestFailed(method, describe(e)))
    try future.get(config.requestTimeout.toMillis, TimeUnit.MILLISECONDS)
    catch
      case _: TimeoutException =>
        future.cancel(true)
        throw BspException(BspError.RequestTimeout(method, config.requestTimeout.toMillis))
      case e: ExecutionException =>
        val cause = if e.getCause != null then e.getCause else e
        throw BspException(BspError.RequestFailed(method, describe(cause)))
      case _: CancellationException =>
        throw BspException(BspError.RequestFailed(method, "request was cancelled"))
      case _: InterruptedException =>
        Thread.currentThread().interrupt()
        future.cancel(true)
        throw BspException(BspError.RequestFailed(method, "interrupted while waiting for response"))

  private def notification(method: String)(call: LsBuildServer => Unit): Unit =
    if closed then throw BspException(BspError.SessionClosed(method))
    try call(server)
    catch case NonFatal(e) => throw BspException(BspError.RequestFailed(method, describe(e)))

  private def describe(t: Throwable): String =
    val cls = t.getClass.getSimpleName
    Option(t.getMessage).filter(_.nonEmpty).map(m => s"$cls: $m").getOrElse(cls)

object BspSession:
  private val threadCounter = new AtomicInteger(1)

  /** Launches the BSP server process described by a connection file (argv as
    * given, cwd = workspace root) and connects over its stdio.
    */
  def launch(
      workspaceRoot: Path,
      details: BspConnectionDetails,
      handlers: BspClientHandlers = BspClientHandlers(),
      config: BspSessionConfig = BspSessionConfig()
  ): BspSession =
    val name = Option(details.getName).getOrElse("<unnamed>")
    val argv = Option(details.getArgv).map(_.asScala.toVector).getOrElse(Vector.empty)
    if argv.isEmpty then
      throw BspException(BspError.LaunchFailed(name, "connection file has empty argv"))
    val process =
      try
        val pb = new ProcessBuilder(argv.asJava)
        pb.directory(workspaceRoot.toFile)
        pb.start()
      catch
        case NonFatal(e) =>
          throw BspException(
            BspError.LaunchFailed(name, Option(e.getMessage).getOrElse(e.getClass.getSimpleName))
          )
    pumpStderr(process, handlers)
    make(workspaceRoot, Some(process), process.getInputStream, process.getOutputStream, handlers, config)

  /** Connects over arbitrary streams (in-process servers in tests, sockets).
    * `input` carries server -> client messages, `output` client -> server.
    */
  def connect(
      workspaceRoot: Path,
      input: InputStream,
      output: OutputStream,
      handlers: BspClientHandlers = BspClientHandlers(),
      config: BspSessionConfig = BspSessionConfig()
  ): BspSession =
    make(workspaceRoot, None, input, output, handlers, config)

  private def make(
      workspaceRoot: Path,
      process: Option[Process],
      input: InputStream,
      output: OutputStream,
      handlers: BspClientHandlers,
      config: BspSessionConfig
  ): BspSession =
    val executor = Executors.newCachedThreadPool(daemonFactory("bsp-session"))
    val client = new ForwardingBuildClient(handlers)
    val launcher: Launcher[LsBuildServer] = new Launcher.Builder[LsBuildServer]()
      .setLocalService(client)
      .setRemoteInterface(classOf[LsBuildServer])
      .setInput(input)
      .setOutput(output)
      .setExecutorService(executor)
      .create()
    val listening = launcher.startListening()
    new BspSession(
      workspaceRoot,
      process,
      input,
      output,
      executor,
      launcher.getRemoteProxy,
      listening,
      config
    )

  private def pumpStderr(process: Process, handlers: BspClientHandlers): Unit =
    val t = new Thread(
      () =>
        try
          val reader =
            new BufferedReader(new InputStreamReader(process.getErrorStream, StandardCharsets.UTF_8))
          var line = reader.readLine()
          while line != null do
            handlers.onServerStderr(line)
            line = reader.readLine()
        catch case NonFatal(_) => (),
      s"bsp-server-stderr-${threadCounter.getAndIncrement()}"
    )
    t.setDaemon(true)
    t.start()

  private def daemonFactory(prefix: String): ThreadFactory = runnable =>
    val t = new Thread(runnable, s"$prefix-${threadCounter.getAndIncrement()}")
    t.setDaemon(true)
    t

File: crates/ls-bsp/src/session.rs
//! One live connection to a BSP server. Every request is bounded by
//! `config.request_timeout` and mapped to a typed [`BspError`]. `launch` starts
//! a server process and connects over its stdio; `connect` runs over arbitrary
//! streams (an in-process server in tests, or a socket).

use std::io::{BufRead, BufReader, Read, Write};
use std::path::PathBuf;
use std::process::{Child, Command, Stdio};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Mutex;
use std::thread;
use std::time::{Duration, Instant};

use serde::de::DeserializeOwned;
use serde_json::Value;

use crate::client::{dispatch_notification, BspClientHandlers};
use crate::discovery::BspConnectionDetails;
use crate::errors::BspError;
use crate::jsonrpc::{JsonRpcClient, Notification, RpcCallError};
use crate::model::BspProjectModel;
use crate::protocol::*;
use crate::uri::path_to_uri;

/// Client identity and the two timeouts governing a session.
#[derive(Clone, Debug)]
pub struct BspSessionConfig {
    pub client_name: String,
    pub client_version: String,
    pub request_timeout: Duration,
    pub shutdown_timeout: Duration,
    /// The waiting-heartbeat interval for the bootstrap-handshake requests
    /// ([`HANDSHAKE_HEARTBEAT_METHODS`]): while such a request has no response
    /// after this interval, one "still waiting for <method>" line is logged per
    /// elapsed interval — the breadcrumb that distinguishes a busy build server
    /// (compiling its build script, or blocked on another mill/sbt holding the
    /// workspace lock) from a wedged one. Tests shrink it to observe the line
    /// quickly; production keeps the 10s default.
    pub handshake_heartbeat: Duration,
}

impl Default for BspSessionConfig {
    fn default() -> Self {
        BspSessionConfig {
            client_name: "scala3-bsp-semantic-ls".to_string(),
            client_version: "0.1.0".to_string(),
            request_timeout: Duration::from_secs(30),
            shutdown_timeout: Duration::from_secs(5),
            handshake_heartbeat: Duration::from_secs(10),
        }
    }
}

/// The bootstrap-handshake requests that get the waiting heartbeat: the ones a
/// restarted server blocks on while the build server starts up (or another
/// build tool holds the workspace lock). `buildTarget/compile` is deliberately
/// NOT here — a long compile is normal, its progress is visible through the
/// forwarded `build/logMessage` lines, and its own begin/end lines bound it.
const HANDSHAKE_HEARTBEAT_METHODS: [&str; 4] = [
    "build/initialize",
    "workspace/buildTargets",
    "buildTarget/sources",
    "buildTarget/scalacOptions",
];

/// Typed result of `buildTarget/compile`.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum BspCompileOutcome {
    Ok {
        origin_id: Option<String>,
    },
    Failed {
        status_code: i32,
        origin_id: Option<String>,
    },
}

impl BspCompileOutcome {
    pub fn is_ok(&self) -> bool {
        matches!(self, BspCompileOutcome::Ok { .. })
    }
}

pub struct BspSession {
    pub workspace_root: PathBuf,
    process: Mutex<Option<Child>>,
    rpc: JsonRpcClient,
    config: BspSessionConfig,
    initialize_result: Mutex<Option<InitializeBuildResult>>,
    closed: AtomicBool,
}

impl BspSession {
    /// Launches the server process described by a connection file (argv as
    /// given, cwd = workspace root) and connects over its stdio.
    pub fn launch(
        workspace_root: PathBuf,
        details: &BspConnectionDetails,
        handlers: BspClientHandlers,
        config: BspSessionConfig,
    ) -> Result<BspSession, BspError> {
        let name = if details.name.is_empty() {
            "<unnamed>".to_string()
        } else {
            details.name.clone()
        };
        if details.argv.is_empty() {
            return Err(BspError::LaunchFailed {
                server: name,
                detail: "connection file has empty argv".to_string(),
            });
        }
        let mut command = Command::new(&details.argv[0]);
        command
            .args(&details.argv[1..])
            .current_dir(&workspace_root)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped());
        let mut child = command.spawn().map_err(|e| BspError::LaunchFailed {
            server: name.clone(),
            detail: e.to_string(),
        })?;
        log::info!(
            target: "bsp",
            "launched build server '{name}' (pid {}), cwd {}",
            child.id(),
            workspace_root.display(),
        );
        let stdout = child.stdout.take().expect("piped stdout");
        let stdin = child.stdin.take().expect("piped stdin");
        let stderr = child.stderr.take().expect("piped stderr");
        pump_stderr(stderr, handlers.clone());
        Ok(Self::make(
            workspace_root,
            Some(child),
            Box::new(stdout),
            Box::new(stdin),
            handlers,
            config,
        ))
    }

    /// Connects over arbitrary streams. `input` carries server -> client
    /// messages, `output` client -> server.
    pub fn connect(
        workspace_root: PathBuf,
        input: Box<dyn Read + Send>,
        output: Box<dyn Write + Send>,
        handlers: BspClientHandlers,
        config: BspSessionConfig,
    ) -> BspSession {
        Self::make(workspace_root, None, input, output, handlers, config)
    }

    fn make(
        workspace_root: PathBuf,
        process: Option<Child>,
        input: Box<dyn Read + Send>,
        output: Box<dyn Write + Send>,
        handlers: BspClientHandlers,
        config: BspSessionConfig,
    ) -> BspSession {
        let dispatch_handlers = handlers.clone();
        let on_notification: Box<dyn Fn(Notification) + Send> =
            Box::new(move |note| dispatch_notification(&dispatch_handlers, note));
        let rpc = JsonRpcClient::start(input, output, on_notification);
        BspSession {
            workspace_root,
            process: Mutex::new(process),
            rpc,
            config,
            initialize_result: Mutex::new(None),
            closed: AtomicBool::new(false),
        }
    }

    pub fn is_closed(&self) -> bool {
        self.closed.load(Ordering::SeqCst)
    }

    /// Alive-flag of the launched server process; None for stream-connected
    /// sessions.
    pub fn server_process_alive(&self) -> Option<bool> {
        let mut guard = self.process.lock().unwrap();
        guard
            .as_mut()
            .map(|child| matches!(child.try_wait(), Ok(None)))
    }

    /// Capabilities from build/initialize; None before [`initialize`] ran.
    pub fn server_capabilities(&self) -> Option<BuildServerCapabilities> {
        self.initialize_result
            .lock()
            .unwrap()
            .as_ref()
            .map(|r| r.capabilities.clone())
    }

    /// build/initialize (languageIds = ["scala"]) then the build/initialized
    /// notification.
    pub fn initialize(&self) -> Result<InitializeBuildResult, BspError> {
        let params = InitializeBuildParams {
            display_name: self.config.client_name.clone(),
            version: self.config.client_version.clone(),
            bsp_version: PROTOCOL_VERSION.to_string(),
            root_uri: path_to_uri(&self.workspace_root),
            capabilities: BuildClientCapabilities {
                language_ids: vec!["scala".to_string()],
            },
        };
        let result: InitializeBuildResult = self.request("build/initialize", to_value(params))?;
        log::info!(
            target: "bsp",
            "build/initialize ok: server '{}' {} (bsp {})",
            result.display_name,
            result.version,
            result.bsp_version,
        );
        *self.initialize_result.lock().unwrap() = Some(result.clone());
        self.notify("build/initialized", Value::Null)?;
        Ok(result)
    }

    pub fn workspace_build_targets(&self) -> Result<Vec<BuildTarget>, BspError> {
        let result: WorkspaceBuildTargetsResult =
            self.request("workspace/buildTargets", Value::Null)?;
        Ok(result.targets)
    }

    pub fn build_target_sources(&self, bsp_ids: &[String]) -> Result<Vec<SourcesItem>, BspError> {
        let params = SourcesParams {
            targets: ids_of(bsp_ids),
        };
        let result: SourcesResult = self.request("buildTarget/sources", to_value(params))?;
        Ok(result.items)
    }

    pub fn build_target_scalac_options(
        &self,
        bsp_ids: &[String],
    ) -> Result<Vec<ScalacOptionsItem>, BspError> {
        let params = ScalacOptionsParams {
            targets: ids_of(bsp_ids),
        };
        let result: ScalacOptionsResult =
            self.request("buildTarget/scalacOptions", to_value(params))?;
        Ok(result.items)
    }

    /// buildTarget/compile. Diagnostics and messages arrive through the client
    /// handlers while the request is in flight; the status code is typed.
    pub fn compile(
        &self,
        bsp_ids: &[String],
        origin_id: Option<String>,
    ) -> Result<BspCompileOutcome, BspError> {
        log::info!(
            target: "bsp",
            "buildTarget/compile started ({} target(s): {:?})",
            bsp_ids.len(),
            bsp_ids,
        );
        let started = Instant::now();
        let params = CompileParams {
            targets: ids_of(bsp_ids),
            origin_id,
        };
        let result: Result<CompileResult, BspError> =
            self.request("buildTarget/compile", to_value(params));
        let elapsed = started.elapsed().as_secs_f64();
        let result = match result {
            Ok(result) => result,
            Err(error) => {
                log::warn!(
                    target: "bsp",
                    "buildTarget/compile failed after {elapsed:.1}s: {error}"
                );
                return Err(error);
            }
        };
        match result.status_code {
            Some(STATUS_OK) => {
                log::info!(
                    target: "bsp",
                    "buildTarget/compile finished: statusCode 1 (OK) in {elapsed:.1}s"
                );
                Ok(BspCompileOutcome::Ok {
                    origin_id: result.origin_id,
                })
            }
            Some(other) => {
                log::warn!(
                    target: "bsp",
                    "buildTarget/compile finished: statusCode {other} in {elapsed:.1}s"
                );
                Ok(BspCompileOutcome::Failed {
                    status_code: other,
                    origin_id: result.origin_id,
                })
            }
            None => Err(BspError::InvalidResponse {
                method: "buildTarget/compile".to_string(),
                detail: "missing statusCode".to_string(),
            }),
        }
    }

    /// Raw buildTarget/inverseSources call, regardless of capabilities.
    pub fn server_inverse_sources(&self, uri: &str) -> Result<Vec<String>, BspError> {
        let params = InverseSourcesParams {
            text_document: TextDocumentIdentifier {
                uri: uri.to_string(),
            },
        };
        let result: InverseSourcesResult =
            self.request("buildTarget/inverseSources", to_value(params))?;
        Ok(result.targets.into_iter().map(|t| t.uri).collect())
    }

    /// Uses the server when it advertises inverseSourcesProvider, otherwise
    /// falls back to the local uri -> target map of the project model.
    pub fn inverse_sources(
        &self,
        uri: &str,
        model: &BspProjectModel,
    ) -> Result<Vec<String>, BspError> {
        let advertised = self
            .server_capabilities()
            .and_then(|caps| caps.inverse_sources_provider)
            .unwrap_or(false);
        if advertised {
            self.server_inverse_sources(uri)
        } else {
            Ok(model.uri_to_target.get(uri).cloned().into_iter().collect())
        }
    }

    /// buildTarget/dependencySources when advertised. Best-effort: not
    /// advertised or a failed request yields None and never crashes the caller.
    pub fn dependency_sources(&self, bsp_ids: &[String]) -> Option<Vec<DependencySourcesItem>> {
        self.capability_gated(
            bsp_ids,
            |caps| caps.dependency_sources_provider,
            || {
                let params = DependencySourcesParams {
                    targets: ids_of(bsp_ids),
                };
                let result: DependencySourcesResult =
                    self.request("buildTarget/dependencySources", to_value(params))?;
                Ok(result.items)
            },
        )
    }

    /// buildTarget/outputPaths when advertised; best-effort like
    /// [`dependency_sources`].
    pub fn output_paths(&self, bsp_ids: &[String]) -> Option<Vec<OutputPathsItem>> {
        self.capability_gated(
            bsp_ids,
            |caps| caps.output_paths_provider,
            || {
                let params = OutputPathsParams {
                    targets: ids_of(bsp_ids),
                };
                let result: OutputPathsResult =
                    self.request("buildTarget/outputPaths", to_value(params))?;
                Ok(result.items)
            },
        )
    }

    fn capability_gated<T>(
        &self,
        bsp_ids: &[String],
        provider: impl Fn(&BuildServerCapabilities) -> Option<bool>,
        call: impl FnOnce() -> Result<Vec<T>, BspError>,
    ) -> Option<Vec<T>> {
        let advertised = self
            .server_capabilities()
            .as_ref()
            .and_then(&provider)
            .unwrap_or(false);
        if bsp_ids.is_empty() || !advertised {
            None
        } else {
            call().ok()
        }
    }

    /// Graceful buildShutdown + build/exit, then stream/process teardown. Each
    /// step is best-effort and bounded by `config.shutdown_timeout`.
    pub fn shutdown(&self) {
        if self.closed.load(Ordering::SeqCst) {
            return;
        }
        log::info!(
            target: "bsp",
            "session shutdown: sending build/shutdown (bounded {:?}) then build/exit",
            self.config.shutdown_timeout,
        );
        let _ = self.request_bounded("build/shutdown", Value::Null, self.config.shutdown_timeout);
        let _ = self.notify("build/exit", Value::Null);
        self.close();
    }

    /// Hard teardown: marks the session closed and terminates the server
    /// process (waiting `config.shutdown_timeout`, then SIGKILL).
    pub fn close(&self) {
        if self.closed.swap(true, Ordering::SeqCst) {
            return;
        }
        // Keep the (now terminated) child in place so `server_process_alive`
        // still reports it after shutdown.
        let mut guard = self.process.lock().unwrap();
        if let Some(child) = guard.as_mut() {
            self.terminate(child);
        }
    }

    fn terminate(&self, child: &mut Child) {
        if wait_quietly(child, self.config.shutdown_timeout) {
            log::info!(target: "bsp", "build server process exited cleanly");
            return;
        }
        // Rust's std only exposes SIGKILL, so the SIGTERM rung of the ladder
        // collapses into the forcible one.
        log::warn!(
            target: "bsp",
            "build server did not exit within {:?} — killing it (SIGKILL)",
            self.config.shutdown_timeout,
        );
        let _ = child.kill();
        wait_quietly(child, Duration::from_millis(1000));
    }

    fn request<T: DeserializeOwned>(&self, method: &str, params: Value) -> Result<T, BspError> {
        let value = self.request_bounded(method, params, self.config.request_timeout)?;
        serde_json::from_value(value).map_err(|e| BspError::InvalidResponse {
            method: method.to_string(),
            detail: e.to_string(),
        })
    }

    fn request_bounded(
        &self,
        method: &str,
        params: Value,
        timeout: Duration,
    ) -> Result<Value, BspError> {
        if self.closed.load(Ordering::SeqCst) {
            return Err(BspError::SessionClosed {
                method: method.to_string(),
            });
        }
        // The bootstrap-handshake requests get the waiting heartbeat; the rest
        // keep the single bounded wait.
        let heartbeat = HANDSHAKE_HEARTBEAT_METHODS
            .contains(&method)
            .then_some(self.config.handshake_heartbeat);
        self.rpc
            .request(method, params, timeout, heartbeat)
            .map_err(|e| match e {
                RpcCallError::Timeout => BspError::RequestTimeout {
                    method: method.to_string(),
                    timeout_millis: timeout.as_millis() as u64,
                },
                RpcCallError::Failed(detail) => BspError::RequestFailed {
                    method: method.to_string(),
                    detail,
                },
            })
    }

    fn notify(&self, method: &str, params: Value) -> Result<(), BspError> {
        if self.closed.load(Ordering::SeqCst) {
            return Err(BspError::SessionClosed {
                method: method.to_string(),
            });
        }
        self.rpc.notify(method, params).map_err(|e| match e {
            RpcCallError::Failed(detail) => BspError::RequestFailed {
                method: method.to_string(),
                detail,
            },
            RpcCallError::Timeout => BspError::RequestTimeout {
                method: method.to_string(),
                timeout_millis: 0,
            },
        })
    }
}

fn ids_of(bsp_ids: &[String]) -> Vec<BuildTargetIdentifier> {
    bsp_ids
        .iter()
        .map(|uri| BuildTargetIdentifier { uri: uri.clone() })
        .collect()
}

fn to_value<T: serde::Serialize>(params: T) -> Value {
    serde_json::to_value(params).unwrap_or(Value::Null)
}

/// Polls the child for up to `timeout`; true if it exited within the window.
fn wait_quietly(child: &mut Child, timeout: Duration) -> bool {
    let deadline = Instant::now() + timeout;
    loop {
        match child.try_wait() {
            Ok(Some(_)) => return true,
            Ok(None) => {
                if Instant::now() >= deadline {
                    return false;
                }
                thread::sleep(Duration::from_millis(10));
            }
            Err(_) => return false,
        }
    }
}

/// Pumps the launched build server's stderr: every line is re-emitted on the
/// log stream under the `bsp-err` area (mill/sbt print their startup and
/// build-script progress here — exactly what a user staring at a silent
/// bootstrap needs to see) and then forwarded to the optional
/// `on_server_stderr` handler (tests capture it there).
fn pump_stderr(stderr: impl Read + Send + 'static, handlers: BspClientHandlers) {
    thread::Builder::new()
        .name("bsp-server-stderr".to_string())
        .spawn(move || {
            let reader = BufReader::new(stderr);
            for line in reader.lines() {
                match line {
                    Ok(line) => {
                        log::info!(target: "bsp-err", "{line}");
                        handlers.emit_stderr(line);
                    }
                    Err(_) => break,
                }
            }
        })
        .ok();
}
